waspc/src/Wasp/GeneratorMonad.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Wasp.GeneratorMonad
+  ( Generator,
+    GeneratorError (..),
+    GeneratorWarning (..),
+    GeneratorState (..),
+    catchGeneratorError,
+    logGeneratorError,
+    logAndThrowGeneratorError,
+    logGeneratorWarning,
+    runGenerator,
+  )
+where
+
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import qualified Control.Monad.Except as MonadExcept
+import Control.Monad.Identity (Identity (runIdentity))
+import Control.Monad.State (MonadState, StateT (runStateT), modify)
+
+-- | Generator is a monad transformer stack where we abstract away the underlying
+-- concrete monad transformers with the helper functions below. This will allow us
+-- to refactor and add more transformers (or swap them) without any caller changes.
+--
+-- The outer Either layer represents the fatal error that halted generation.
+-- There may optionally be other errors or warnings logged in the State.
+newtype Generator a = Generator
+  { _runGenerator :: ExceptT GeneratorError (StateT GeneratorState Identity) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState GeneratorState,
+      MonadError GeneratorError
+    )
+
+data GeneratorState = GeneratorState
+  { warnings :: [GeneratorWarning],
+    errors :: [GeneratorError]
+  }
+
+data GeneratorError
+  = GenericGeneratorError String
+  | PackageJsonGenerationError String
+  | PrismaGenerationError String
+
+instance Show GeneratorError where
+  show (GenericGeneratorError e) = e
+  show (PackageJsonGenerationError e) = "package.json Generation Error: " ++ e
+  show (PrismaGenerationError e) = "Prisma Generation Error: " ++ e
+
+data GeneratorWarning = GenericGeneratorWarning String
+
+instance Show GeneratorWarning where
+  show (GenericGeneratorWarning e) = e
+
+runGenerator :: Generator a -> (Either GeneratorError a, GeneratorState)
+runGenerator generator = runIdentity $ runStateT (runExceptT (_runGenerator generator)) state
+  where
+    state = GeneratorState {warnings = [], errors = []}
+
+logGeneratorWarning :: GeneratorWarning -> Generator ()
+logGeneratorWarning w = modify $ \GeneratorState {errors = errors', warnings = warnings'} ->
+  GeneratorState {errors = errors', warnings = w : warnings'}
+
+logGeneratorError :: GeneratorError -> Generator ()
```

Comment:
I think this goes back to the original idea of "Does the presence of _any_ error imply failure". Originally I was thinking no, which is why I added, but since we are going with Yes there for our Generator return type, I am inclined to remove this to force clients along that path. Sound good?

Reply:
Sure, sounds good! I mean, we do want to let Generator catch errors -> so for example if DbGenerator fails with an error and throws it, we might want to catch that error in the top lvl Generator, so that we can also call other generator subsystems (server, client, ...), in order to collect more errors so that we can report them all at once. However, although we caught that error, we still want to consider the whole "generation" as failed. The only purpose of cathing an error is to continue computation and collect more of errors, not to actually recover from it (at least not for now, and I doubt we will ever do that).

So with that in mind, logGeneratorError would be equivalent to logAndThrowGeneratorError that we immediately caught at that very same -> I am not sure if that is something we will want to do, mostly likely not, and if we will want to do it, we can easily support it in the future.

TLDR; yes, let's make it also throw!

Reply:
Cool sounds good, I will at least nix it for now until we see a use case for the immediate catch. 👍🏻

Reply:
Moved that function as a helper into `logAndThrowGeneratorError` to ensure they can only do that for now, and know what is happening:

```haskell
-- This logs an error and does throw, thus short-circuiting the computation unless caught.
logAndThrowGeneratorError :: GeneratorError -> Generator a
logAndThrowGeneratorError e = logGeneratorError >> throwError e
  where
    logGeneratorError :: Generator ()
    logGeneratorError = modify $ \GeneratorState {errors = errors', warnings = warnings'} ->
      GeneratorState {errors = e : errors', warnings = warnings'}
```